src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -111,9 +144,211 @@ internal class DiagnosticComputer
+                //    to minimize resource and CPU contention between normal priority tasks
+                //    and the high priority computeTask in Step 4 below.
+                CancelNormalPriorityTasks();
+
+                // Step 4:
+                //  - Execute the core 'computeTask' for diagnostic computation.
+                return await computeTask.ConfigureAwait(false);
+            }
+            finally
+            {
+                // Step 5:
+                //  - Remove the 'computeTask' from the list of current executing high priority tasks.
+                lock (s_gate)
+                {
+                    var removed = s_highPriorityComputeTasks.Remove(computeTask);
+                    Debug.Assert(removed);
+                }
+            }
+
+            static void CancelNormalPriorityTasks()
+            {
+                ImmutableArray<CancellationTokenSource> cancellationTokenSourcesToCancel;
+                lock (s_gate)
+                {
+                    cancellationTokenSourcesToCancel = s_cancellationSourcesForNormalPriorityComputeTasks.ToImmutableArrayOrEmpty();
+                }
+
+                foreach (var cancellationTokenSource in cancellationTokenSourcesToCancel)
+                {
+                    try
+                    {
+                        cancellationTokenSource.Cancel();
+                    }
+                    catch (ObjectDisposedException)
+                    {
+                        // CancellationTokenSource might get disposed if the normal priority
+                        // task completes while we were executing this foreach loop.
+                        // Gracefully handle this case and ignore this exception.
+                    }
+                }
+            }
+        }
+
+        private async Task<SerializableDiagnosticAnalysisResults> GetNormalPriorityDiagnosticsAsync(
+            IEnumerable<string> analyzerIds,
+            bool reportSuppressedDiagnostics,
+            bool logPerformanceInfo,
+            bool getTelemetryInfo,
+            CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // Step 1:
+                //  - Normal priority task must wait for all the executing high priority tasks to complete
+                //    before beginning execution.
+                await WaitForHighPriorityTasksAsync(cancellationToken).ConfigureAwait(false);
+
+                // Step 2:
+                //  - Create the core 'computeTask' for computing diagnostics.
+                //  - Create a custom 'cancellationTokenSource' associated with this 'computeTask'.
+                //    This token source allows normal priority computeTasks to be cancelled when
+                //    a subsequent high priority diagnostic request is received.
+                using var cancellationTokenSource = new CancellationTokenSource();
```

Comment:
fwiw, give that we already have a try/finally below... consider not placing this in a using, but instead just clean it up with the rest of hte cleanup.  another benefit of this is that we can hten push creation of this CTS into 'CreateComputeTask'

Reply:
I explicitly wanted to keep this CTS creation in this top level method to ensure clear scope and lifetime. We would also need to introduce a new try/finally OR move steps 2 and 3 inside the existing try block if we wanted to cleanup this cts in the finally block. It just felt more logical to have a `using var` here instead.